File: src/NodeParser/LiteralNodeParser.ts
import ts from "typescript";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";

export class LiteralNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.LiteralTypeNode): boolean {
        return node.kind === ts.SyntaxKind.LiteralType;
    }
    public createType(node: ts.LiteralTypeNode, context: Context): BaseType {
        return this.childNodeParser.createType(node.literal, context);
    }
}

File: internal/parser/literal_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// LiteralNodeParser handles literal type nodes by delegating to their inner
// literal (src/NodeParser/LiteralNodeParser.ts).
type LiteralNodeParser struct {
	childNodeParser NodeParser
}

func NewLiteralNodeParser(childNodeParser NodeParser) *LiteralNodeParser {
	return &LiteralNodeParser{childNodeParser: childNodeParser}
}

func (p *LiteralNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindLiteralType
}

func (p *LiteralNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return p.childNodeParser.CreateType(node.AsLiteralTypeNode().Literal, ctx, nil)
}
